### DISCUSSION TITLE
How to build Tasmota with Sensor and Display

### DISCUSSION THREAD

**Discussion by smougenot** [arendst/Tasmota#11254] [Q&A] [answered]
## Goal

Give my kid a [TTGO Display](https://github.com/smougenot/TTGO-T-Display) that shows time and current temperature (BMP280).
![pinmap](https://user-images.githubusercontent.com/1217669/110247965-10823c00-7f6f-11eb-87c3-b434591febe1.jpg)

So I need both "abilities" of Tasmota
* I2C sensors 
* Display

I already coded a [custom firmware in a github repo](https://github.com/smougenot/esp32_tft_sensors) but it is far from what Tasmota can do.

## Attempts 

So far I could not build the firmware

I tried to combine the settings of the pio environments but this not ended well
The code base is the `v9.3.1` version (tag in git)
```
[env:tasmota32-sensor-display]
extends                 = env:tasmota32
build_flags             = ${common32.build_flags} -DFIRMWARE_SENSORS -DFIRMWARE_DISPLAYS
lib_extra_dirs          = lib/libesp32, lib/lib_basic, lib/lib_display, lib/lib_i2c, lib/lib_rf, lib/lib_div, lib/lib_ssl
```
![2021-03-07_17-57](https://user-images.githubusercontent.com/1217669/110247919-df097080-7f6e-11eb-84a6-209b050a81ac.png)

## Help
Can someone help achieving this?

**Comment by barbudor** [comment, marked as answer]
Create a custom env `[env:tasmota32-sensor-display]` is a good way
But you shouldn't use predefined FIRMWARE_SENSORS FIRMWARE_DISPLAYS because they will enable and disable of features that will conflict
Read again : https://tasmota.github.io/docs/Compile-your-build/ and especially the last section on creating your own custom firmware

1) Your env should be:
```
[env:tasmota32-sensor-display]
extends                 = env:tasmota32
build_flags             = ${common32.build_flags} -DMY_FW_SENSORS_DISPLAYS
```
This should be in `platformio_tasmota_cenv.ini` (never modify a standard tasmota file)

2) In your `user_config_override.h`
```
#ifdef MY_FW_SENSORS_DISPLAYS
    // -- CODE_IMAGE_STR is the name shown between brackets on the 
    //    Information page or in INFO MQTT messages
    #undef  :CODE_IMAGE_STR
    #define CODE_IMAGE_STR "sensorsdisplays"

   // Copy here all defines that you need to support your display. 
   // You can copy lines from tasmota_configurations.h in FIRMWARE_DISPLAYS such as 
   #define USE_DISPLAY                              // Add Display Support (+2k code)
   #define USE_SPI                                  // Hardware SPI using GPIO12(MISO), GPIO13(MOSI) and GPIO14(CLK) in addition to two user selectable GPIOs(CS and DC)
   #define USE_DISPLAY_ST7789                     // [DisplayModel 12] Enable ST7789 module

   // Copy here all defines that you need to support your sensors. 
   // You can copy lines from tasmota_configurations.h in FIRMWARE_SENSORS such as 
   #define USE_I2C                                  // I2C using library wire (+10k code, 0k2 mem, 124 iram)
   #define USE_BMP                                // [I2cDriver10] Enable BMP085/BMP180/BMP280/BME280 sensors (I2C addresses 0x76 and 0x77) (+4k4 code)

#endif // #ifdef MY_FW_SENSORS_DISPLAYS
```

Bonne chance 👍

**Reply by smougenot** [reply]
Thank you very much for the help
I'll try this ASAP.
And maybe build a PR to share the result.

**Reply by smougenot** [reply]
Sorry to ask help again @barbudor but it seams to me the **Display features are not working** : 
* the command `Display` is unknown 
* even using the predefined environment for the **display firmware**
* might be because of my device (I do not know if commands are device dependant ... rather than firmware dependant) ?

### I tried several branches 
1. `development`
1. `master` (the screen capture)
![2021-03-09_23-59](https://user-images.githubusercontent.com/1217669/110550156-761f2580-8133-11eb-8c49-f91ced971b74.png)

### Process to test
1. Get clean code from Github on the tested branches
2. Change (remove from 2 ini files) pio setting `upload_port`
2. run command `pio run -t upload && pio device monitor --baud 115200` (with the device connected)

### I could 
* set the Wifi, 
* trigger ordinary commands (`State`, `I2CDriver`, ...)

### Please help 
what am I missing?

**Reply by smougenot** [reply]
As a last recourse, I also tried the pre-build firmware at http://ota.tasmota.com/tasmota32/release/tasmota32-display.bin
With the same result : Display answer `Command Unknown`

**Reply by barbudor** [reply]
Bonjour Sylvain,

Have you configured the pins for your display ?
Until Tasmota is confirmed to have a display attached, the display features remains disabled, even if in the binary
You need to properly configure the GPIO:
- SPI MOSI
- SPI CLK
- ST7789 DC
- ST7789 CS
- OLED Reset (optional I think)

**Reply by smougenot** [reply]
I missed it
Thank you for the tip, I did not understood that.
I'll give it a try ASAP

Merci beaucoup

**Reply by Jason2866** [reply]
Have you configured the display GPIOs? Without the commands will not get active. The commands only work when the display is recognized from Tasmota. Take a look in the console when Tasmota starts

**Reply by smougenot** [reply]
Thank you very much for the help,
It now respond to the `Display` command

Yet nothing is displayed but, I'll continue to investigate.

I hope I'll get far enough to post here the detailled solution to use Tasmota with this device and use the display.

**Comment by Jason2866** [comment]
The error you encountered is solved with latest development version.
There was a conflict between the TSL2561 and the epaper42 driver.
Still everything barbudor has written is correct

**Reply by smougenot** [reply]
Thank you for the information.
Honestly I did not dug into the error myself (because I expect conflicts on untested combinations).

**Reply by Jason2866** [reply]
We try to test many combinations. Feedback about non working ones are very appreciated

**Comment by smougenot** [comment]
:tada: I managed to do it
* I'll soon post a detailed description here for anyone who wanna start using that kind of settings

**Reply by barbudor** [reply]
May be  write a new 'how to article' in the Show and Tell section would be  a great idea

**Reply by smougenot** [reply]
:+1: 
I'll give it a try, not sure I'm the most qualified 
but still glad if it helps
